utils/connects.stanza

```diff
@@ -12,41 +12,89 @@ defpackage ocdb/connects :
   import ocdb/bundles
   import ocdb/symbols
   import ocdb/box-symbol
+  import lang-utils
 
-public defn connect-isolate (a:JITXObject, b:JITXObject) :
-  inside pcb-module :
-    eval-when true :
-      match(port-type(a), port-type(b)) :
-        (pa:Bundle, pb:Bundle) :
-          if pa == i2c and pb == i2c :
-            inst buf : ocdb/texas-instruments/ISO1540/component
-            net (a.sda, buf.SDA1)
-            net (a.scl, buf.SCL1)
-            net (b.sda, buf.SDA2)
-            net (b.scl, buf.SCL2)
-            property(buf.VCC1.requires-power) = true
-            property(buf.VCC1.power-request) = [3.3 0.3 0.05]
+; Helper function to check if the ports have bundles and their assigned
+; bundle matches what we expect
+defn ensure-bundles!<?T> (
+  body: () -> ?T, 
+  ports:Seqable<JITXObject>, 
+  bundles:Seqable<Bundle>,
+  action) -> False|T:
+  ; collect the mismatched bundles from ports
+  val errors = 
+    for (p in ports, b in bundles) seq?:
+      match(port-type(p)):
+        (b?:Bundle):
+          if b? != b:
+            One([p, b, b?])  ; case 1: ports exist but do not match
+          else:
+            None()           ; case 2: ports exist and match, we're good to go
+        (_:?):
+          One([p, b, false]) ; case 3: port is not a bundle
+  
+  ; error if not empty
+  if not empty?(errors):
+    val buf = StringBuffer()
+    println(buf, "Could not %_. Bundles did not match." % [action])
+    for [p, b, b?] in errors do:
+      println(IndentedStream(buf), 
+        "Expected %_ for port %_. Found %_." % [
+          name(b), ref(p), 
+          name(b? as Bundle) when b? is-not False else false
+        ]
+      )
+    fatal(to-string(buf)) ; todo error or exception here?
```

Comment:
I would make a public `BundleNotFound` (or a more apt name) exception in connections.stanza
This should be an exception since it occurs when a designer provides a bad object, and not because of a connections implementation error.

Reply:
I feel like that falls under "blame the caller." The error handling in ocdb is already up in the air - I don't think we throw exceptions anywhere internally in this library. But I'm not sure what our error handling story should look like

Reply:
Agreed. Either we have an `ocdb/errors.stanza` or inline the Exception to this file.